src/sasl2_fast.js

```diff
@@ -0,0 +1,239 @@
+            }
+        }
+    },
+
+    /**
+     * @param {Element} elem 
+     */
+    onAuth: function (elem) {
+
+        // Note: this looks *a lot* like sasl2.onAuth(),
+        // but it runs on a different tag, a sub-tag of the main <authentication> stanza
+
+        const server_mechanisms = new Set([...elem.querySelectorAll('mechanism')]
+            .map((m) => m.textContent));
+
+        this.mechanisms = [...server_mechanisms
+            .intersection(
+                new Set(Object.keys(this.conn.mechanisms))
+            )];
+        // sort by priority
+        this.mechanisms.sort((a, b) => {
+            (this.conn.mechanisms[b].priority - this.conn.mechanisms[a].priority)
+        })
+
+        console.info(
+            "FAST: server advertised ", server_mechanisms,
+            " of which we support", this.mechanisms);
+        if (this.mechanisms.length == 0) {
+            console.warn("FAST offered but with no known mechanisms.");
+            return
+        }
+
+        // Append to the list of SASL2 mechanisms
+        // XXX what happens if this runs before sasl2.onAuth()?
+        this.conn.sasl2.mechanisms = [...
+            new Set(this.conn.sasl2.mechanism)
+                .union(new Set(this.mechanisms))
+        ];
+        // sort by priority XXX the FAST mechanisms should have higher priority!
+        this.conn.sasl2.mechanisms.sort((a, b) => {
+            return (this.conn.mechanisms[b].priority - this.conn.mechanisms[a].priority)
+        })
+
+        return false; // stop listening
+    },
+
+    /**
+     * @param {Element} elem 
+     */
+    onToken: function (elem) {
+
+        this.credential = {
+            ...this.credential,
+            'token': elem.getAttribute('token'),
+            'expiry': Date.parse(elem.getAttribute('expiry')),
+            'counter': 0,
+        };
+        console.log("fast plugin onToken", elem, this.credential)
+
+        return true; // keep listening: the server is allowed to rotate our token anytime it wishes
+    },
+
+    logout: async function () {
+        // Invalidate the FAST token on log out
+        // XXX this does not seem to actually get sent,
+        // and Converse does not forget the token from its IndexedDB
+        // if you edit Local Storage using the web debugger to re-add conversejs-session-jid: 'user@xmpp.example.org'
+        // and reload then FAST will happily log you back in
+
+        if (this.credential.token) {
+            let authenticate = await this.conn.sasl2.authenticateStanza(this.credential.mechanism)
+
+            // XXX copy-pasta
+            const response = await this.conn.mechanisms[this.credential.mechanism].clientChallenge(this.conn);
+            authenticate
+                .c('initial-response',
+                    null,
+                    btoa(/** @type {string} */(response)))
+                .up();
+
+            authenticate.nodeTree.querySelector("fast")?.setAttribute("invalidate", "true")
```

Comment:
This whole thing doesn't work. In https://github.com/conversejs/converse.js/pull/3693 I made sure the token gets forgotten on logout, so it's not the _worst_, but it would be ideal to invalidate tokens we're not using.